README.md
![](assets/headimage.png)

<p align="center">
    <a href="https://pypi.org/project/efpodsanalyzer/">
        <img src="https://badge.fury.io/py/efpodsanalyzer.svg">
    </a>
    <a href="https://www.python.org/">
        <img src="https://img.shields.io/badge/language-Python-2e6fa0.svg">
    </a>
    <a href="https://codebeat.co/projects/github-com-eyrefree-efpodsanalyzer-master">
        <img src="https://codebeat.co/badges/67a3cc17-24fb-4c3d-b94c-61e17eea08cc"/>
    </a>
    <a href="https://raw.githubusercontent.com/EyreFree/EFPodsAnalyzer/master/LICENSE">
        <img src="https://img.shields.io/badge/license-GPLv3-000000.svg">
    </a>
    <a href="https://twitter.com/EyreFree777">
        <img src="https://img.shields.io/badge/twitter-@EyreFree777-blue.svg?style=flat">
    </a>
    <a href="http://weibo.com/eyrefree777">
        <img src="https://img.shields.io/badge/weibo-@EyreFree-red.svg?style=flat">
    </a>
    <a href="https://raw.githubusercontent.com/EyreFree/EFQRCode/assets/icon/MadeWith%3C3.png">
        <img src="https://img.shields.io/badge/made%20with-%3C3-orange.svg">
    </a>
</p>

EFPodsAnalyzer is a Python script that is used to help us sort out the CoaoaPods library dependency relationship. It can generate clear Pods dependency graph with only one command, which will help us in dependency sorting / component cleaning.

> [中文介绍](/README_CN.md)

## Preview

The more the Pods library is, the more complex the library is, the more it is used to get the tool, for example, the modularized item; conversely, if the project dependency is very few and the dependency relation is very simple, there is no need for this tool. The dependency graph of a project of our company is as follows, looks like a shit:

![](assets/overview.png)

## Requirements

- Python 2.7

## Installation

### PyPI

If you have [pip](https://github.com/pypa/pip) installed on your device, you can install the latest version of this tool directly with the following commands:

```
sudo pip install efpodsanalyzer --upgrade
```

### Manual

Download this project, or execute the following commands to Clone this project:

```
git clone git@github.com:EyreFree/EFPodsAnalyzer.git
```

## Usage

### PyPI

1. First, ensure that your project has been operated on `pod install` and successfully generated the Pods directory;
2. Install `efpodsanalyzer` on your device with `pip`;
3. Classify the dependent libraries according to the specific circumstances of the projects to be analyzed, and give the regex of each category separately, then write the rules to the `EFPAConfig.json` file under the directory which contains the target `Podfile`;
4. Execute the following commands for the generation of the dependency graph:

```
sudo efpodsanalyzer [Target Podfile file path]
```

5. You can check the output of the terminal. If you see the following log, you can open the generated file `index.html` in the path with your browser:

```
Dependency graph generated: .../EFPADiagram/index.html
```

If there is any error, please deal with the corresponding error information, or you can make a PR or an Issue.

### Manual

1. First, ensure that your project has been operated on `pod install` and successfully generated the Pods directory;
2. The content of this tool has been fully downloaded to the local area;
3. Classify the dependent libraries according to the specific circumstances of the projects to be analyzed, and give the regex of each category separately, then write the rules to the `EFPAConfig.json` file under the directory which contains the target `Podfile`;
4. Execute the following commands for the generation of the dependency graph:

```
python [EFPodsAnalyzer.py file path] [Target Podfile file path]
```

5. You can check the output of the terminal. If you see the following log, you can open the generated file `index.html` in the path with your browser:

```
Dependency graph generated: .../EFPADiagram/index.html
```

If there is any error, please deal with the corresponding error information, or you can make a PR or an Issue.

## Example

Here we take [Coding's open source iOS client](https://github.com/Coding/Coding-iOS) as an example to show you the complete use of this tool:

1. First, install `efpodsanalyzer` on your device with `pip`.
2. Then download the Coding iOS project to local, and perform `pod install` operation;
3. Because the dependency of the Coding client is basically not classified, which is all the third party library. Therefore, for demonstration purposes, I divide it into three categories: the library with the beginning of 'M', the library ending with 'Kit' and the others. The `EFPAConfig.json` is as follows(You can also [click here](/efpodsanalyzer/EFPAConfig.json) to see the sample file online):

```
{
    "config": {
        "categories": ["M-prefixed Libraries", "Library ending with Kit", "Other"],
        "categoryRegexes": ["^M.*", ".*(Kit)$", ".*"]
    }
}
```

4. The command to generate diagram is as follows:

```
sudo efpodsanalyzer /Users/eyrefree/Documents/iOS_GitHub/Coding-iOS/Podfile
```

5. The final dependency graph is generated as follows, you can also [view it online](https://eyrefree.github.io/EFPodsAnalyzer/docs/graph_force.html):

![](assets/example.png)

- Clicking on the top class name can control the display and hiding of the class;
- The mouse can be suspended on the node to display the node name;
- The mouse can be suspended on the line between two nodes to display the relationship of them.

Another style `circular`, as follows, you can [view it online](https://eyrefree.github.io/EFPodsAnalyzer/docs/graph_circular.html) too:

![](assets/example_circular.png)

## Todo

- Determine whether the dependency needs to be removed, according to the header file reference;
- Determine whether the dependency needs to be removed, depending on the class dependency;
- More style of diagrams.

## Other

1. The view of dependency graph is based on [ECharts](https://github.com/ecomfe/echarts) and [xml2json](https://github.com/abdmob/x2js), thanks for their work!
2. Code of this tool can be packaged / released using the following commands. People who needs to build a custom version can make some self exploration(Under the root directory of this project):

 ```
rm -rf dist/*;
python setup.py sdist bdist_wheel;
twine upload dist/efpodsanalyzer*;
```

## Author

EyreFree, eyrefree@eyrefree.org

## License

![](https://www.gnu.org/graphics/gplv3-127x51.png)

EFPodsAnalyzer is available under the GPLv3 license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log

-----

## [0.2.6](https://github.com/EyreFree/EFPodsAnalyzer/releases/tag/0.2.6) (2021-09-12)

* Fix issue [#5](https://github.com/EyreFree/EFPodsAnalyzer/issues/5).

---

## [0.2.5](https://github.com/EyreFree/EFPodsAnalyzer/releases/tag/0.2.5) (2021-05-20)

#### Fix

* Fix dead loop problem caused by sub dependency without first level reference relationship.

---

## [0.2.4](https://github.com/EyreFree/EFPodsAnalyzer/releases/tag/0.2.4) (2020-09-11)

#### Fix

* Fix `include_package_data`.

---

## [0.2.3](https://github.com/EyreFree/EFPodsAnalyzer/releases/tag/0.2.3) (2020-09-11)

#### Fix

* Fix `ValueError: must have exactly one of create/read/write/append mode`.

---

## [0.2.2](https://github.com/EyreFree/EFPodsAnalyzer/releases/tag/0.2.2) (2020-09-11)

#### Add

* Add default `EFPAConfig.json` if not exist.

---

## [0.2.1](https://github.com/EyreFree/EFPodsAnalyzer/releases/tag/0.2.1) (2019-02-10)

#### Fix

* Fix `style.css`.

---

## [0.2.0](https://github.com/EyreFree/EFPodsAnalyzer/releases/tag/0.2.0) (2019-02-10)

#### Add

* Add reference count;
* Add new style `circular`.

---

## [0.1.4](https://github.com/EyreFree/EFPodsAnalyzer/releases/tag/0.1.4) (2019-02-07)

#### Add

* Wrap with PyPI.

---

## [0.0.2](https://github.com/EyreFree/EFPodsAnalyzer/releases/tag/0.0.2) (2018-09-12)

#### Fix

* Fix dead cycle.

---

## [0.0.1](https://github.com/EyreFree/EFPodsAnalyzer/releases/tag/0.0.1) (2018-02-09)

First public release.


EFPodsAnalyzer.py
# !/usr/bin/python
# -*- coding: UTF-8 -*-

from __future__ import print_function

import json
import os
import re
import sys
import webbrowser

# 错误码
ERROR_PARAMETERS_COUNT = 1  # 参数数量错误
ERROR_PARAMETERS_TYPE = 2   # 参数错误
ERROR_NEED_PODS = 3         # Pods 目录不存在或不完整
ERROR_FILE_CORRUPTED = 4    # 文件损坏或格式不正确

# 文件名长度
SELF_FILENAME_LEN = len(sys.argv[0].split('/')[-1])

# 原始数据单元


class ManifestPodClass:
    'Manifest中取出的未处理数据单元'
    podName = ''
    podDependencies = []

    def __init__(self, name, dependencies):
        self.podName = name
        self.podDependencies = dependencies

    def dependenciesCount(self):
        return len(self.podDependencies)

    @staticmethod
    def printList(Objectlist):
        for object in Objectlist:
            print(object.podName, object.podDependencies)

# 最终数据单元


class PodClass:
    '处理后的数据单元'
    podName = ''
    podDependencyIndexes = []

    def __init__(self, name, dependencyIndexes):
        self.podName = name
        self.podDependencyIndexes = dependencyIndexes

    def dependenciesCount(self):
        return len(self.podDependencyIndexes)

    def dependencies(self, podList):
        return [podList[i] for i in self.podDependencyIndexes]

    @staticmethod
    def printList(Objectlist):
        for object in Objectlist:
            print(object.podName, object.podDependencyIndexes)

# 从 Manifest.lock 读取 PODS 原始数据单元列表


def readManifestPodListFromFile(fileName):
    fo = open(fileName, "r+")
    manifestFileContent = fo.read()
    fo.close()
    spliteList = manifestFileContent.split('DEPENDENCIES:')
    if len(spliteList) <= 1:
        print('Error: The Manifest.lock file is corrupted!')
        exit(ERROR_FILE_CORRUPTED)
    podsString = '  ' + spliteList[0][5:].strip(' \n')
    podsStringList = podsString.split('\n')
    # 生成原始数据单元数组
    returnList = []
    index = -1
    lastObject = ManifestPodClass('', [])
    for line in podsStringList:
        lineClean = line.replace('\'','').replace('\"','')
        podName = lineClean.lstrip(' -').split(' ')[0]
        if lineClean.startswith('    - '):
            if index >= 0:
                lastObject.podDependencies.append(podName)
        elif lineClean.startswith('  - '):
            if index >= 0:
                returnList.append(lastObject)
            index = index + 1
            lastObject = ManifestPodClass(podName, [])
        else:
            print('Error: The Manifest.lock file is corrupted!')
            exit(ERROR_FILE_CORRUPTED)
    returnList.append(lastObject)
    return returnList

# 从 ManifestPodList 生成最终数据单元列表


def generatePodListFromList(manifestPodList):
    def getBaseIndexes(dependencies, baseList):
        returnList = []
        if 0 == len(dependencies):
            return returnList
        for dependency in dependencies:
            tempMark = False
            for index, base in enumerate(baseList):
                if base.podName == dependency:
                    returnList.append(index)
                    tempMark = True
            if False == tempMark:
                return returnList
        return returnList

    returnList = []
    nextList = list(manifestPodList)
    while 0 < len(nextList):
        doingList = list(nextList)
        for manifestPod in doingList:
            baseIndexes = getBaseIndexes(
                manifestPod.podDependencies, returnList)
            if len(baseIndexes) == len(manifestPod.podDependencies):
                returnList.append(PodClass(manifestPod.podName, baseIndexes))
                nextList.remove(manifestPod)
    return returnList

# 生成依赖关系图


def generateDependencyGraph(podlist):
    # Setting
    cc = open(sys.argv[0][:-SELF_FILENAME_LEN] + "config.json", "r+")
    configString = cc.read()
    cc.close()

    configObject = json.loads(configString)

    settingTitle = sys.argv[1][:-8].split('/')[-1]
    settingCategories = json.dumps(configObject['config']['categories'])
    settingCategoryRegexes = configObject['config']['categoryRegexes']
    settingContent = "setting = '{\"setting\": {\"title\": \"" + \
        settingTitle + "\",\"categories\": " + str(settingCategories) + "}}';"
    xx = open(sys.argv[0][:-SELF_FILENAME_LEN] + "docs/setting.json", "wr+")
    xx.write(settingContent)
    xx.close()

    # Data
    nodeString = ""
    for index, pod in enumerate(podlist):
        podID = str(index)
        podLabel = pod.podName
        podAttvalue = str(1)
        for index, regex in enumerate(settingCategoryRegexes):
            if re.match(regex, pod.podName):
                # print(pod.podName + " -> " + regex)
                podAttvalue = str(index)
                break

        nodeString = nodeString + "{\"attvalues\": {\"attvalue\": {\"_for\": \"modularity_class\",\"_value\": \"" + podAttvalue + \
            "\"}},\"size\": {\"_value\": \"24\",\"__prefix\": \"viz\"},\"position\": {\"_x\": \"-225.73984\",\"_y\": \"82.41631\",\"_z\": \"0.0\",\"__prefix\": \"viz\"},\"color\": {\"_r\": \"236\",\"_g\": \"81\",\"_b\": \"72\",\"__prefix\": \"viz\"},\"_id\": \"" + podID + "\",\"_label\": \"" + podLabel + "\"}"
        nodeString = nodeString + ","

    if nodeString.endswith(","):
        nodeString = nodeString[:-1]

    edgeString = ""
    edgeIndex = 0
    for index, pod in enumerate(podlist):
        if pod.dependenciesCount() > 0:
            for podDependencyIndex in pod.podDependencyIndexes:
                podID = str(edgeIndex)
                podSource = str(index)
                podTarget = str(podDependencyIndex)
                edgeIndex += 1
                edgeString = edgeString + "{\"attvalues\": \"\",\"_id\": \"" + podID + "\",\"_source\": \"" + \
                    podSource + "\",\"_target\": \"" + podTarget + "\",\"_weight\": \"17.0\"}"
                edgeString = edgeString + ","

    if edgeString.endswith(","):
        edgeString = edgeString[:-1]

    oo = open(sys.argv[0][:-SELF_FILENAME_LEN] + "template/data.json", "r+")
    dataTemplate = oo.read()
    oo.close()

    dataContent = "data = \'" + dataTemplate.replace("$nodes$", nodeString).replace(
        "$edges$", edgeString).replace("\n", "") + "\';"
    xx = open(sys.argv[0][:-SELF_FILENAME_LEN] + "docs/data.json", "wr+")
    xx.write(dataContent)
    xx.close()

    graphHtmlPath = sys.argv[0][:-SELF_FILENAME_LEN] + "docs/index.html"
    print("Dependency graph generated: " + graphHtmlPath)

    webbrowser.open("file://" + graphHtmlPath)

    return


# 输入参数数量错误直接返回错误
if 2 != len(sys.argv):
    print('Error：Count of parameters is not correct!')
    exit(ERROR_PARAMETERS_COUNT)

# 输入文件不是 Podfile 直接返回错误
if True != sys.argv[1].endswith('Podfile'):
    print('Error: Illegal parameters!')
    exit(ERROR_PARAMETERS_TYPE)

# 判断 Pods 目录下的 Manifest.lock 是否存在，否则返回错误
manifestFileName = sys.argv[1][:-4] + 's/Manifest.lock'
if True != os.path.exists(manifestFileName):
    print('Error: Please `pod install` first!')
    exit(ERROR_NEED_PODS)

# 读取 PODS 依赖关系并生成依赖关系数组
manifestPodlist = readManifestPodListFromFile(manifestFileName)
# ManifestPodClass.printList(manifestPodlist)
podlist = generatePodListFromList(manifestPodlist)
# PodClass.printList(podlist)

# 生成依赖关系页面
generateDependencyGraph(podlist)


EFImportChecker.py
# !/usr/bin/python
# -*- coding: UTF-8 -*-

from __future__ import print_function

import json
import os
import re
import sys
import webbrowser

# 错误码
ERROR_PARAMETERS_COUNT = 1  # 参数数量错误
ERROR_PARAMETERS_TYPE = 2   # 参数错误
ERROR_NEED_PODS = 3         # Pods 目录不存在或不完整
ERROR_FILE_CORRUPTED = 4    # 文件损坏或格式不正确

# 文件名长度
SELF_FILENAME_LEN = len(sys.argv[0].split('/')[-1])

# 原始数据单元


class ManifestPodClass:
    'Manifest中取出的未处理数据单元'
    podName = ''
    podDependencies = []

    def __init__(self, name, dependencies):
        self.podName = name
        self.podDependencies = dependencies

    def dependenciesCount(self):
        return len(self.podDependencies)

    @staticmethod
    def printList(Objectlist):
        for object in Objectlist:
            print(object.podName, object.podDependencies)

# 最终数据单元


class PodClass:
    '处理后的数据单元'
    podName = ''
    podDependencyIndexes = []

    def __init__(self, name, dependencyIndexes):
        self.podName = name
        self.podDependencyIndexes = dependencyIndexes

    def dependenciesCount(self):
        return len(self.podDependencyIndexes)

    def dependencies(self, podList):
        return [podList[i] for i in self.podDependencyIndexes]

    @staticmethod
    def printList(Objectlist):
        for object in Objectlist:
            print(object.podName, object.podDependencyIndexes)

# 从 Manifest.lock 读取 PODS 原始数据单元列表


def readManifestPodListFromFile(fileName):
    fo = open(fileName, "r+")
    manifestFileContent = fo.read()
    fo.close()
    spliteList = manifestFileContent.split('DEPENDENCIES:')
    if len(spliteList) <= 1:
        print('Error: The Manifest.lock file is corrupted!')
        exit(ERROR_FILE_CORRUPTED)
    podsString = '  ' + spliteList[0][5:].strip(' \n')
    podsStringList = podsString.split('\n')
    # 生成原始数据单元数组
    returnList = []
    index = -1
    lastObject = ManifestPodClass('', [])
    for line in podsStringList:
        podName = line.lstrip(' -').split(' ')[0]
        if line.startswith('    - '):
            if index >= 0:
                lastObject.podDependencies.append(podName)
        elif line.startswith('  - '):
            if index >= 0:
                returnList.append(lastObject)
            index = index + 1
            lastObject = ManifestPodClass(podName, [])
        else:
            print('Error: The Manifest.lock file is corrupted!')
            exit(ERROR_FILE_CORRUPTED)
    returnList.append(lastObject)
    return returnList

# 从 ManifestPodList 生成最终数据单元列表


def generatePodListFromList(manifestPodList):
    def getBaseIndexes(dependencies, baseList):
        returnList = []
        if 0 == len(dependencies):
            return returnList
        for dependency in dependencies:
            tempMark = False
            for index, base in enumerate(baseList):
                if base.podName == dependency:
                    returnList.append(index)
                    tempMark = True
            if False == tempMark:
                return returnList
        return returnList

    returnList = []
    nextList = list(manifestPodList)
    while 0 < len(nextList):
        doingList = list(nextList)
        for manifestPod in doingList:
            baseIndexes = getBaseIndexes(
                manifestPod.podDependencies, returnList)
            if len(baseIndexes) == len(manifestPod.podDependencies):
                returnList.append(PodClass(manifestPod.podName, baseIndexes))
                nextList.remove(manifestPod)
    return returnList

# 生成依赖关系图


def generateDependencyGraph(podlist):
    # Setting
    cc = open(sys.argv[0][:-SELF_FILENAME_LEN] + "config.json", "r+")
    configString = cc.read()
    cc.close()

    configObject = json.loads(configString)

    settingTitle = sys.argv[1][:-8].split('/')[-1]
    settingCategories = json.dumps(configObject['config']['categories'])
    settingCategoryRegexes = configObject['config']['categoryRegexes']
    settingContent = "setting = '{\"setting\": {\"title\": \"" + \
        settingTitle + "\",\"categories\": " + str(settingCategories) + "}}';"
    xx = open(sys.argv[0][:-SELF_FILENAME_LEN] + "docs/setting.json", "wr+")
    xx.write(settingContent)
    xx.close()

    # Data
    nodeString = ""
    for index, pod in enumerate(podlist):
        podID = str(index)
        podLabel = pod.podName
        podAttvalue = str(1)
        for index, regex in enumerate(settingCategoryRegexes):
            if re.match(regex, pod.podName):
                # print(pod.podName + " -> " + regex)
                podAttvalue = str(index)
                break

        nodeString = nodeString + "{\"attvalues\": {\"attvalue\": {\"_for\": \"modularity_class\",\"_value\": \"" + podAttvalue + \
            "\"}},\"size\": {\"_value\": \"24\",\"__prefix\": \"viz\"},\"position\": {\"_x\": \"-225.73984\",\"_y\": \"82.41631\",\"_z\": \"0.0\",\"__prefix\": \"viz\"},\"color\": {\"_r\": \"236\",\"_g\": \"81\",\"_b\": \"72\",\"__prefix\": \"viz\"},\"_id\": \"" + podID + "\",\"_label\": \"" + podLabel + "\"}"
        nodeString = nodeString + ","

    if nodeString.endswith(","):
        nodeString = nodeString[:-1]

    edgeString = ""
    edgeIndex = 0
    for index, pod in enumerate(podlist):
        if pod.dependenciesCount() > 0:
            for podDependencyIndex in pod.podDependencyIndexes:
                podID = str(edgeIndex)
                podSource = str(index)
                podTarget = str(podDependencyIndex)
                edgeIndex += 1
                edgeString = edgeString + "{\"attvalues\": \"\",\"_id\": \"" + podID + "\",\"_source\": \"" + \
                    podSource + "\",\"_target\": \"" + podTarget + "\",\"_weight\": \"17.0\"}"
                edgeString = edgeString + ","

    if edgeString.endswith(","):
        edgeString = edgeString[:-1]

    oo = open(sys.argv[0][:-SELF_FILENAME_LEN] + "template/data.json", "r+")
    dataTemplate = oo.read()
    oo.close()

    dataContent = "data = \'" + dataTemplate.replace("$nodes$", nodeString).replace(
        "$edges$", edgeString).replace("\n", "") + "\';"
    xx = open(sys.argv[0][:-SELF_FILENAME_LEN] + "docs/data.json", "wr+")
    xx.write(dataContent)
    xx.close()

    graphHtmlPath = sys.argv[0][:-SELF_FILENAME_LEN] + "docs/index.html"
    print("Dependency graph generated: " + graphHtmlPath)

    webbrowser.open("file://" + graphHtmlPath)

    return


# 输入参数数量错误直接返回错误
if 2 != len(sys.argv):
    print('Error：Count of parameters is not correct!')
    exit(ERROR_PARAMETERS_COUNT)

# 输入文件不是 Podfile 直接返回错误
if True != sys.argv[1].endswith('Podfile'):
    print('Error: Illegal parameters!')
    exit(ERROR_PARAMETERS_TYPE)

# 判断 Pods 目录下的 Manifest.lock 是否存在，否则返回错误
manifestFileName = sys.argv[1][:-4] + 's/Manifest.lock'
if True != os.path.exists(manifestFileName):
    print('Error: Please `pod install` first!')
    exit(ERROR_NEED_PODS)

# 读取 PODS 依赖关系并生成依赖关系数组
manifestPodlist = readManifestPodListFromFile(manifestFileName)
# ManifestPodClass.printList(manifestPodlist)
podlist = generatePodListFromList(manifestPodlist)
# PodClass.printList(podlist)

# 生成依赖关系页面
generateDependencyGraph(podlist)


efpodsanalyzer/efpodsanalyzer.py
# !/usr/bin/python
# -*- coding: UTF-8 -*-

from __future__ import print_function

import json
import os
import re
import sys
# import webbrowser
import shutil

# 错误码
ERROR_PARAMETERS_COUNT = 1  # 参数数量错误
ERROR_PARAMETERS_TYPE = 2   # 参数错误
ERROR_NEED_PODS = 3         # Pods 目录不存在或不完整
ERROR_FILE_CORRUPTED = 4    # 文件损坏或格式不正确
ERROR_FILE_NOT_EXIST = 5    # 文件不存在

# 文件名长度
SELF_FILENAME_LEN = len(sys.argv[0].split('/')[-1])
POD_FILENAME_LEN = len(sys.argv[1].split('/')[-1])

# 路径相关
PROJECT_NAME = "efpodsanalyzer"
MODULE_NAME = PROJECT_NAME.lower()
ROOT_PROJECT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
MODULE_DIR = os.path.join(ROOT_PROJECT_DIR, MODULE_NAME)

# 原始数据单元
class ManifestPodClass:
    'Manifest中取出的未处理数据单元'
    podName = ''
    podDependencies = []

    def __init__(self, name, dependencies):
        self.podName = name
        self.podDependencies = dependencies

    def dependenciesCount(self):
        return len(self.podDependencies)

    @staticmethod
    def printList(Objectlist):
        for object in Objectlist:
            print(object.podName, object.podDependencies)

# 最终数据单元
class PodClass:
    '处理后的数据单元'
    podName = ''
    podDependencyIndexes = []
    podReferenceCount = 1

    def __init__(self, name, dependencyIndexes):
        self.podName = name
        self.podDependencyIndexes = dependencyIndexes

    def dependenciesCount(self):
        return len(self.podDependencyIndexes)

    def dependencies(self, podList):
        return [podList[i] for i in self.podDependencyIndexes]

    @staticmethod
    def printList(Objectlist):
        for object in Objectlist:
            print(object.podName, object.podDependencyIndexes)

# 从 Manifest.lock 读取 PODS 原始数据单元列表
def readManifestPodListFromFile(fileName):
    fo = open(fileName, "r+")
    manifestFileContent = fo.read()
    fo.close()
    spliteList = manifestFileContent.split('DEPENDENCIES:')
    if len(spliteList) <= 1:
        print('Error: The Manifest.lock file is corrupted!')
        exit(ERROR_FILE_CORRUPTED)
    podsString = '  ' + spliteList[0][5:].strip(' \n')
    podsStringList = podsString.split('\n')
    # 生成原始数据单元数组
    returnList = []
    index = -1
    lastObject = ManifestPodClass('', [])
    for line in podsStringList:
        lineClean = line.replace('\'','').replace('\"','')
        podName = lineClean.lstrip(' -').split(' ')[0]
        if lineClean.startswith('    - '):
            if index >= 0:
                lastObject.podDependencies.append(podName)
        elif lineClean.startswith('  - '):
            if index >= 0:
                returnList.append(lastObject)
            index = index + 1
            lastObject = ManifestPodClass(podName, [])
        else:
            print('Error: The Manifest.lock file is corrupted!')
            exit(ERROR_FILE_CORRUPTED)
    returnList.append(lastObject)
    return returnList

# 从 ManifestPodList 生成最终数据单元列表
def generatePodListFromList(manifestPodList):
    def getBaseIndexes(dependencies, baseList):
        returnList = []
        if 0 == len(dependencies):
            return returnList
        for dependency in dependencies:
            #print("dependency: " + dependency)
            tempMark = False
            for index, base in enumerate(baseList):
                #print("dependency: " + dependency + ", base.podName: " + base.podName)
                if base.podName == dependency:
                    returnList.append(index)
                    tempMark = True
            if False == tempMark:
                return returnList
        return returnList

    returnList = []
    nextList = list(manifestPodList)
    while 0 < len(nextList):
        doingList = list(nextList)
        for manifestPod in doingList:
            baseIndexes = getBaseIndexes(manifestPod.podDependencies, returnList)
            if len(baseIndexes) == len(manifestPod.podDependencies):
                returnList.append(PodClass(manifestPod.podName, baseIndexes))
                nextList.remove(manifestPod)
                # 权重
                for baseIndex in baseIndexes:
                    returnList[baseIndex].podReferenceCount += 1
                    # print(str(baseIndex) + "+= 1, podReferenceCount = " + str(returnList[baseIndex].podReferenceCount))
            else:
                # 解决一种奇怪的特殊情况
                for podDependencie in manifestPod.podDependencies:
                    checkSpecial = True
                    for rtnItem in returnList:
                        if rtnItem.podName == podDependencie:
                            checkSpecial = False
                    for nextItem in nextList:
                        if nextItem.podName == podDependencie:
                            checkSpecial = False
                    if checkSpecial == True:
                        returnList.append(PodClass(podDependencie, []))
    return returnList

# 生成依赖关系图
def generateDependencyGraph(podlist, configString):
    # Directory
    directory = podFilePath() + "EFPADiagram"
    if not os.path.exists(directory):
        os.makedirs(directory)

    configObject = json.loads(configString)

    settingTitle = podFileDirectoryName()
    settingCategories = json.dumps(configObject['config']['categories'])
    settingCategoryRegexes = configObject['config']['categoryRegexes']
    settingContent = "setting = '{\"setting\": {\"title\": \"" + \
        settingTitle + "\",\"categories\": " + str(settingCategories) + "}}';"
    xx = open(podFilePath() + "EFPADiagram/setting.json", "w+")
    xx.write(settingContent)
    xx.close()

    # Data
    nodeString = ""
    for index, pod in enumerate(podlist):
        podValue = str(pod.podReferenceCount)
        podID = str(index)
        podLabel = pod.podName
        podAttvalue = str(1)
        for index, regex in enumerate(settingCategoryRegexes):
            if re.match(regex, pod.podName):
                # print(pod.podName + " -> " + regex)
                podAttvalue = str(index)
                break

        nodeString = nodeString + "{\"attvalues\": {\"attvalue\": {\"_for\": \"modularity_class\",\"_value\": \"" + podAttvalue + \
            "\"}},\"size\": {\"_value\": \"" + podValue + "\",\"__prefix\": \"viz\"},\"position\": {\"_x\": \"-225.73984\",\"_" + \
            "y\": \"82.41631\",\"_z\": \"0.0\",\"__prefix\": \"viz\"},\"color\": {\"_r\": \"236\",\"_g\": \"81\",\"_b\": \"72\"" + \
            ",\"__prefix\": \"viz\"},\"_id\": \"" + podID + "\",\"_label\": \"" + podLabel + "\"}"
        nodeString = nodeString + ","

    if nodeString.endswith(","):
        nodeString = nodeString[:-1]

    edgeString = ""
    edgeIndex = 0
    for index, pod in enumerate(podlist):
        if pod.dependenciesCount() > 0:
            for podDependencyIndex in pod.podDependencyIndexes:
                podID = str(edgeIndex)
                podSource = str(index)
                podTarget = str(podDependencyIndex)
                edgeIndex += 1
                edgeString = edgeString + "{\"attvalues\": \"\",\"_id\": \"" + podID + "\",\"_source\": \"" + \
                    podSource + "\",\"_target\": \"" + podTarget + "\",\"_weight\": \"16.0\"}"
                edgeString = edgeString + ","

    if edgeString.endswith(","):
        edgeString = edgeString[:-1]

    oo = open(resourcePath("template/data.json"), "r+")
    dataTemplate = oo.read()
    oo.close()

    dataContent = "data = \'" + dataTemplate.replace("$nodes$", nodeString).replace("$edges$", edgeString).replace("\n", "") + "\';"
    xx = open(podFilePath() + "EFPADiagram/data.json", "w+")
    xx.write(dataContent)
    xx.close()

    # JS
    directoryJS = podFilePath() + "EFPADiagram/js"
    if not os.path.exists(directoryJS):
        os.makedirs(directoryJS)
    shutil.copy(resourcePath("EFPADiagram/js/dataTool.js"), directoryJS + "/dataTool.js")
    shutil.copy(resourcePath("EFPADiagram/js/echarts.min.js"), directoryJS + "/echarts.min.js")
    shutil.copy(resourcePath("EFPADiagram/js/jquery.min.js"), directoryJS + "/jquery.min.js")
    shutil.copy(resourcePath("EFPADiagram/js/xml2json.min.js"), directoryJS + "/xml2json.min.js")

    # CSS
    directoryCSS = podFilePath() + "EFPADiagram/css"
    if not os.path.exists(directoryCSS):
        os.makedirs(directoryCSS)
    shutil.copy(resourcePath("EFPADiagram/css/style.css"), directoryCSS + "/style.css")

    # HTML
    graphHtmlPath = podFilePath() + "EFPADiagram/"
    shutil.copy(resourcePath("EFPADiagram/graph_circular.html"), graphHtmlPath + "graph_circular.html")
    shutil.copy(resourcePath("EFPADiagram/graph_force.html"), graphHtmlPath + "graph_force.html")
    shutil.copy(resourcePath("EFPADiagram/index.html"), graphHtmlPath + "index.html")
    print("Dependency graph generated: " + graphHtmlPath)

    # webbrowser.open("file://" + graphHtmlPath)

    return

# 生成资源文件目录访问路径
def resourcePath(relativePath):
    return os.path.join(MODULE_DIR, relativePath)

# 当前文件所在路径
def selfFilePath():
    return sys.argv[0][:-SELF_FILENAME_LEN]

# 目标文件所在路径
def podFilePath():
    return sys.argv[1][:-POD_FILENAME_LEN]

# 目标文件上级目录名
def podFileDirectoryName():
    return sys.argv[1][:-8].split('/')[-1]

# 主要
def main():
    # 输入参数数量错误直接返回错误
    if 2 != len(sys.argv):
        print('Error：Count of parameters is not correct!')
        exit(ERROR_PARAMETERS_COUNT)

    # 输入文件不是 Podfile 直接返回错误
    if True != sys.argv[1].endswith('Podfile'):
        print('Error: Illegal parameters!')
        exit(ERROR_PARAMETERS_TYPE)

    # 判断 Pods 目录下的 Manifest.lock 是否存在，否则返回错误
    manifestFileName = sys.argv[1][:-4] + 's/Manifest.lock'
    if True != os.path.exists(manifestFileName):
        print('Error: Please run `pod install` first!')
        exit(ERROR_NEED_PODS)

    # 判断目标目录下的 EFPAConfig.json 是否存在，如果不存在的话，则使用默认的配置文件
    configFileName = podFilePath() + '/EFPAConfig.json'
    configString = '''
    {
        "config": {
            "categories": ["Other"],
            "categoryRegexes": [".*"]
        }
    }
    '''
    if True == os.path.exists(configFileName):
        cc = open(podFilePath() + "EFPAConfig.json", "r+")
        configString = cc.read()
        cc.close()

    # 读取 PODS 依赖关系并生成依赖关系数组
    manifestPodlist = readManifestPodListFromFile(manifestFileName)
    # ManifestPodClass.printList(manifestPodlist)
    podlist = generatePodListFromList(manifestPodlist)
    # PodClass.printList(podlist)

    # 生成依赖关系页面
    generateDependencyGraph(podlist, configString)

# 执行
if __name__ == "__main__":
    main()
